lib/Sema/CSApply.cpp

```diff
@@ -3594,13 +3631,68 @@ namespace {
       llvm_unreachable("Unhandled OverloadChoiceKind in switch.");
     }
 
+    /// Builds an argument list for making a call to the given
+    /// `@dynamicMemberLookup` subscript by inserting any necessary default
+    /// arguments.
+    ArgumentList *buildArgumentListForDynamicMemberLookupSubscript(
+        const SelectedOverload &overload, SourceLoc componentLoc,
+        ConstraintLocator *locator) {
+      assert(overload.choice.isAnyDynamicMemberLookup() &&
+             "Overload must be for dynamic member lookup");
+      auto *SD = cast<SubscriptDecl>(overload.choice.getDecl());
+      auto memberLoc = cs.getCalleeLocator(locator);
+      auto subscript = resolveConcreteDeclRef(SD, memberLoc);
+
+      auto openedFullFnTy =
+          simplifyType(overload.adjustedOpenedFullType)->castTo<FunctionType>();
+      auto fullSubscriptTy =
+          openedFullFnTy->getResult()->castTo<FunctionType>();
+      auto params = fullSubscriptTy->getParams();
+
+      SmallVector<Expr *, 4> args;
+      for (auto paramIdx : indices(params)) {
+        Expr *argExpr;
+        if (paramIdx == 0) {
+          auto paramTy = params[paramIdx].getPlainType();
+          if (overload.choice.isKeyPathDynamicMemberLookup()) {
+            argExpr = buildKeyPathDynamicMemberArgExpr(paramTy, componentLoc,
+                                                       memberLoc);
+          } else {
+            auto fieldName =
+                overload.choice.getName().getBaseIdentifier().str();
+            argExpr = buildDynamicMemberLookupArgExpr(fieldName, componentLoc,
+                                                      paramTy);
+          }
+        } else {
+          // If bindings were substituted, we need to find the "original" (or
+          // contextless) parameter index for the default argument.
+          if (shouldSubstituteParameterBindings(subscript)) {
+            auto *paramList = SD->getParameterList();
+            assert(paramList);
+            paramIdx = paramList->getOrigParamIndex(
+                subscript.getSubstitutions(), paramIdx);
+          }
+
+          auto owner = getDefaultArgOwner(subscript, paramIdx);
+          auto paramTy = params[paramIdx].getParameterType();
+          argExpr = new (ctx)
+              DefaultArgumentExpr(owner, paramIdx, componentLoc, paramTy, dc);
```

Comment:
This is pretty much what happens in `coerceCallArguments`, but trying to DRY this out didn't end up feeling worth it.